### PR TITLE
Add flask async view

### PR DESCRIPTION
**Pull request by scottweitzner**
## Description

MR to address issue https://github.com/strawberry-graphql/strawberry/issues/1903

Adds a new async-enabled view for use with Flask. This is specifically helpful when using async dataloaders to handle n+1 cases in your schema

## Types of Changes
- [ ] Core
- [ ] Bugfix
- [x] New feature
- [x] Enhancement/optimization
- [ ] Documentation

## Issues Fixed or Closed by This PR

https://github.com/strawberry-graphql/strawberry/issues/1903

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by botberry** [comment]
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Scott Weitzner for the PR 👏

Get it here 👉 https://github.com/strawberry-graphql/strawberry/releases/tag/(next)
```

File: strawberry/flask/views.py [hunk 1]
```diff
@@ -92,3 +92,55 @@ def dispatch_request(self) -> Response:
...
+        except MissingQueryError:
+            return Response("No valid query was provided for the request", 400)
+
+        response = Response(status=200, content_type="application/json")
+        context = self.get_context(response)
+
+        allowed_operation_types = OperationType.from_http(method)
+
+        if not self.allow_queries_via_get and method == "GET":
+            allowed_operation_types = allowed_operation_types - {OperationType.QUERY}
```
Comment (scottweitzner) [inline strawberry/flask/views.py:129]: @patrick91 I'm super conscious of the fact that this is an exact copy of the logic in the existing view. 

As a potential solution to this we could overload the dispatch_request to be optionally async. I have found [this](https://stackoverflow.com/questions/39778901/optional-synchronous-interface-to-asynchronous-functions) as a guide to be a potential solution for this, however it has some noticeable defects. 

I'm happy to restructure this code to whatever will be desirable to you.

Comment (patrick91) [reply strawberry/flask/views.py:129]: no worries at all! we'd like to refactor __all__ the views to reduce the code duplication at some point. Hopefully we find a solution for this, but even if we don't we won't have as much duplicated code as we have now :)

I've done some precursory work here (just for the tests), if you're curious: https://github.com/strawberry-graphql/strawberry/pull/1840

File: RELEASE.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+Release type: patch
```
Comment (patrick91) [inline RELEASE.md:1]:
```suggestion
Release type: minor
```

let's go with minor as this is basically a new feature 😊

File: RELEASE.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+This release adds a new flask view to allow for aysn dispatching of requests.
```
Comment (patrick91) [inline RELEASE.md:3]:
```suggestion
This release adds a new flask view to allow for aysnc dispatching of requests.
```

File: strawberry/flask/views.py [hunk 4]
```diff
@@ -92,3 +92,55 @@ def dispatch_request(self) -> Response:
...
+    async def dispatch_request(self):
+        method = request.method
+        content_type = request.content_type or ""
+
+        if "application/json" in content_type:
+            data: dict = request.json  # type:ignore[assignment]
+        elif content_type.startswith("multipart/form-data"):
+            operations = json.loads(request.form.get("operations", "{}"))
+            files_map = json.loads(request.form.get("map", "{}"))
+
```
Comment (patrick91) [inline strawberry/flask/views.py:107]: missing coverage here is fine, it should be improved by my PR: https://github.com/strawberry-graphql/strawberry/pull/1840 
so no worries about this

**Review by patrick91 — COMMENTED** [review]
Looks pretty good! Can you add some docs here? https://strawberry.rocks/docs/integrations/flask

then it will be ready to merge 😊

**Comment by scottweitzner** [comment]
@patrick91 Added a few more tests to get that coverage up and also added the docs. Thank you for making this an enjoyable process to contribute back. Looking forward to making more contributions down the line

File: docs/integrations/flask.md [hunk 5]
```diff
@@ -24,6 +24,12 @@ if __name__ == "__main__":
     app.run()
 ```
 
+If you'd prefer to use an asynchronous view you can instead use the following import which has the same interface as `GraphQLView`. This is helpful if using a dataloader.
+
+```python
+from strawberry.flask.views import GraphQLView
```
Comment (estyxx) [inline docs/integrations/flask.md:30]:
```suggestion
from strawberry.flask.views import AsyncGraphQLView
```
Do you mean `AsyncGraphQLView` maybe?

**Review by estyxx — COMMENTED** [review]
Thank you for this PR! I added a couple of comments :)

**Review by patrick91 — APPROVED** [review]
Looks great! I took the liberty of applying @estyxx suggestions 😊 will merge in a bit :)

**Comment by botberry** [comment]
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

File: docs/integrations/flask.md (hunk 5 above)
Comment (scottweitzner) [reply docs/integrations/flask.md]: Good catch. Thank you!

**Comment by scottweitzner** [comment]
> Looks great! I took the liberty of applying @estyxx suggestions 😊 will merge in a bit :)

Thanks all!

**Pull request merged**